### PR TITLE
NIFI-11294 - Add support for component state as checkpointing strategy

### PR DESCRIPTION
**Pull request by malthe**
# Summary

[NIFI-11294](https://issues.apache.org/jira/browse/NIFI-11294) ConsumeAzureEventHub should default to processor state for checkpointing

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [ ] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [ ] Build completed using `mvn clean install -P contrib-check`
  - [ ] JDK 11
  - [x] JDK 17

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [x] Documentation formatting appears as expected in rendered files

### CODE REVIEW

**Comment by malthe** [comment]
Note that the existing tests (which I have adapted, somewhat unsuccessfully) are not very good – they do not actually exercise the processor's logic end-to-end, but only the "meat". I think the whole test setup for this processor really needs to be reworked entirely.

**Comment by turcsanyip** [comment]
Will review...

**Review by exceptionfactory — COMMENTED** [review]
Thanks for implementing this feature @malthe! Just noting that the automated builds flagged a couple Checkstyle violations, and the unit tests failed due to unnecessary stubbing. That usually means there are unnecessary mock calls that need to be removed or adjusted so the `when` expectations are only called when needed.

**Comment by malthe** [comment]
@exceptionfactory thanks – these issues have been fixed now.

The good is that the actual code seems pretty well-functioning and reasonable well-crafted as well. However, the tests are bad/wrong and I fear I have only made them worse.

**Comment by malthe** [comment]
@turcsanyip I have now implemented proper load-balancing in a9bd297b50515918092ec01a0730ee795e917932 – and to fix the issues with stopping the processor, I have removed most of the processor state fields, shifting them into a event batch processing closure instead.

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/eventhub/ConsumeAzureEventHub.java [hunk 1]
```diff
@@ -486,11 +563,36 @@ protected String getTransitUri(final PartitionContext partitionContext) {
...
+                                new ComponentStateCheckpointStore.State() {
+                                    public StateMap getState() throws IOException {
+                                        return session.getState(Scope.CLUSTER);
+                                    }
+
+                                    public boolean replaceState(StateMap oldValue, Map<String, String> newValue) {
+                                        return false;
+                                    }
+                                }
+                        ),
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/eventhub/ConsumeAzureEventHub.java:586]: The inner anonymous class results in multiple levels of nesting that make this hard to follow. Recommend breaking it out to an explicit class.

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/eventhub/ConsumeAzureEventHub.java [hunk 2]
```diff
@@ -403,36 +420,76 @@ public void stopClient() {
...
+        if (checkpointStrategy == CheckpointStrategy.AZURE_BLOB_STORAGE) {
+            final String containerName = defaultIfBlank(context.getProperty(STORAGE_CONTAINER_NAME).evaluateAttributeExpressions().getValue(), eventHubName);
+            final String storageConnectionString = createStorageConnectionString(context);
+            final BlobContainerAsyncClient blobContainerAsyncClient = getBlobContainerAsyncClient(containerName, storageConnectionString);
+            checkpointStore = getCheckpointStoreFromBlobContainer(blobContainerAsyncClient);
+            legacyPartitionEventPosition = getLegacyPartitionEventPosition(blobContainerAsyncClient, consumerGroup);
+        } else {
+            checkpointStore = new ComponentStateCheckpointStore(
+                    identifier,
+                    new ComponentStateCheckpointStore.State() {
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/eventhub/ConsumeAzureEventHub.java:447]: Having two different anonymous inner class implementations make this hard to follow. Recommend breaking this out to a distinct class.

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for the updates on this new feature @malthe! I noted several test implementation concerns, and a couple recommendations on the implementation itself.

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/eventhub/utils/ComponentStateCheckpointStore.java [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+                            KEY_OWNERSHIP,
+                            partitionOwnership.getFullyQualifiedNamespace(),
+                            partitionOwnership.getEventHubName(),
+                            partitionOwnership.getConsumerGroup(),
+                            partitionOwnership.getPartitionId()
+                    );
+
+                    final Map<String, String> newState = new HashMap<>(oldState.toMap());
+                    long timestamp = System.currentTimeMillis();
+                    String eTag = identifier + "/" + timestamp;
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/eventhub/utils/ComponentStateCheckpointStore.java:92]: Should the `eTag` be created based on the state version as opposed the current timestamp? That seems like it would align better with the purpose of the tag for tracking ownership and ensuring consistent version control across nodes.

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/eventhub/utils/ComponentStateCheckpointStore.java [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+            final String fullyQualifiedNamespace = parts[1];
+            final String eventHubName = parts[2];
+            final String consumerGroup = parts[3];
+            final String partitionId = parts[4];
+            PartitionContext partitionContext = new PartitionContext(
+                    fullyQualifiedNamespace,
+                    eventHubName,
+                    consumerGroup,
+                    partitionId
+            );
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/eventhub/utils/ComponentStateCheckpointStore.java:211]: Recommend pulling this logic out to a separate method so the formatting approach is clear.

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/eventhub/ConsumeAzureEventHub.java [hunk 5]
```diff
@@ -403,36 +420,76 @@ public void stopClient() {
...
+                        public StateMap getState() throws IOException {
+                            final ProcessSession session = processSessionFactory.createSession();
+                            return session.getState(Scope.CLUSTER);
+                        }
+
+                        public boolean replaceState(StateMap oldValue, Map<String, String> newValue) throws IOException {
+                            final ProcessSession session = processSessionFactory.createSession();
+                            if (!session.replaceState(oldValue, newValue, Scope.CLUSTER)) {
```
Comment (turcsanyip) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/eventhub/ConsumeAzureEventHub.java:455]: `ProcessSession.replaceState()` does not provide fully proper optimistic locking because the state can be changed by another session between `replaceState()` and `commit()`. So even if `replaceState()` returns `true`, the state change may be omitted at commit time due to a concurrent update.

I suggest using `ProcessContext.replace()` instead which has the right optimistic locking semantics which is required from a checkpoint store implementation.

Please note `ProcessContext.replace()` cannot initialize the state currently (see also [NIFI-11595](https://issues.apache.org/jira/browse/NIFI-11595)) so it needs to be created with `setState()` before using `replace()`. E.g. in `@OnScheduled` like this:
```
    @OnScheduled
    public void onScheduled(ProcessContext context) throws IOException {
        if (getNodeTypeProvider().isPrimary()) {
            final StateManager stateManager = context.getStateManager();
            final StateMap state = stateManager.getState(Scope.CLUSTER);

            if (!state.getStateVersion().isPresent()) {
                stateManager.setState(new HashMap<>(), Scope.CLUSTER);
            }
        }
    }
```

**Review by turcsanyip — CHANGES_REQUESTED** [review]
@malthe I tried the processor but it seems to process all messages again and again after restarts. I cannot see the checkpoint data saved in the state (only ownership) so I think this is the root cause. Could you please check it?

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/test/java/org/apache/nifi/processors/azure/eventhub/TestConsumeAzureEventHub.java [hunk 6]
```diff
@@ -117,13 +131,22 @@ public class TestConsumeAzureEventHub {
...
+    private ArgumentCaptor<Consumer<EventBatchContext>> eventBatchProcessorCapture;
+
     private MockConsumeAzureEventHub processor;
 
     private TestRunner testRunner;
 
     @BeforeEach
     public void setupProcessor() {
-        processor = new MockConsumeAzureEventHub();
+        processor = spy(new MockConsumeAzureEventHub());
```
[the remark this answers is not in the document]
Comment (malthe) [reply nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/test/java/org/apache/nifi/processors/azure/eventhub/TestConsumeAzureEventHub.java:149]: Previously, the processor wasn't tested through the `onTrigger` but through some internal methods.

I changed the tests so that all testing happens through the actually exposed interface. However, this requires some additional mocking in order to implement the checkpoint strategy for example.

(The reason being, we don't have the luxury of testing against a real event hub here.)

But instead of spying on the processor, I suppose I could implement some alternative methods on the mock class. I don't remember if I tried that and faced some issues perhaps.

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/test/java/org/apache/nifi/processors/azure/eventhub/utils/TestComponentStateCheckpointStore.java [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+    public void testClaimOwnershipSuccess() {
+        var claimed = checkpointStore.claimOwnership(List.of(
+            partitionOwnership
+        )).blockFirst();
+        var listed = checkpointStore.listOwnership(
+                EVENT_HUB_NAMESPACE,
+                EVENT_HUB_NAME,
+                CONSUMER_GROUP
+        ).blockFirst();
+        assert listed.getETag().equals(claimed.getETag());
```
[the remark this answers is not in the document]
Comment (malthe) [reply nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/test/java/org/apache/nifi/processors/azure/eventhub/utils/TestComponentStateCheckpointStore.java]: Fixed in 4aa8240c0ed8dc52dedcd7e4463e1672d5abf34d.

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/eventhub/utils/ComponentStateCheckpointStore.java (hunk 3 above)
Comment (malthe) [reply nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/eventhub/utils/ComponentStateCheckpointStore.java:92]: I'm not sure – but that logic was copied from the blob-based checkpoint store.

In fact, whenever I could, I have used the exact same logic between them.

**Comment by turcsanyip** [comment]
@malthe Just a heads-up that [NIFI-11595](https://issues.apache.org/jira/browse/NIFI-11595) has been merged fixing the `replaceState()` logic I mentioned in [my previous comment](https://github.com/apache/nifi/pull/7076#discussion_r1206003566). It makes the state handling of the checkpoints simpler, though it was not a blocker.

Considering that there was no activity on the PR in the past months, I would like to ask if you are still interested in this feature? Do you intend to finish the PR? I'm happy to jump in and continue the implementation based on your PR if you have other interests now.

Thanks

**Comment by malthe** [comment]
@turcsanyip please do – I am currently not actively using NiFi so just following along from the sideline.

**Comment by turcsanyip** [comment]
Thanks @malthe! I'll go ahead and try to finish the implementation.

**Comment by exceptionfactory** [comment]
Thanks for your work on this @malthe, and thanks for planning to take this on @turcsanyip. I am closing the pull request for now given the state of several conflicts, but feel free to open a new one when it is ready for review.
